test/Dapr.E2E.Test.App/Actors/ReentrantActor.cs

```diff
@@ -0,0 +1,65 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Dapr.Actors.Runtime;
+
+namespace Dapr.E2E.Test.Actors.Reentrancy
+{
+    public class ReentrantActor : Actor, IReentrantActor
+    {
+        public ReentrantActor(ActorHost host)
+            : base(host)
+        {
+        }
+
+        public Task Ping()
+        {
+            return Task.CompletedTask;
+        }
+
+        // An actor method that exercises reentrancy by calling more methods in the same actor.
+        // Can be configured to different reentrant depths via the ReentrantCallOptions but will
+        // always make at least one additional call.
+        public async Task ReentrantCall(ReentrantCallOptions callOptions)
+        {
+            try
+            {
+                await UpdateState(true, callOptions.CallNumber);
+                var actor = this.ProxyFactory.CreateActorProxy<IReentrantActor>(this.Id, "ReentrantActor");
+                if (callOptions == null || callOptions.CallsRemaining <= 1)
+                {   
+                    await actor.Ping();
+                }
+                else
+                {
+                    await actor.ReentrantCall(new ReentrantCallOptions 
+                    { 
+                        CallsRemaining = callOptions.CallsRemaining - 1,
+                        CallNumber = callOptions.CallNumber + 1,
+                    });
+                }                
+            } 
+            finally 
```

Comment:
Do we need the finally here? I'm asking because we discussed this pattern and said we'd tell users to avoid it.

Reply:
Nah, I can move that method out.

Out of curiosity, why do we tell customers to avoid that pattern?

Reply:
That pattern means that you commit the result even when a failure happens. It leads to potential concurrency issues accessing the state manager - we largely fixed within the state manager, but user code is not likely to be thread-safe.